backvendor/gosource.go

```diff
@@ -395,15 +400,32 @@ func findImportComment(src *GoSource) (string, error) {
 			return filepath.SkipDir
 		}
 
-		pkg, err := build.ImportDir(pth, build.ImportComment)
+		pkg, err := build.ImportDir(path, build.ImportComment)
 		if err != nil {
 			if _, ok := err.(*build.NoGoError); ok {
 				return nil
 			}
 			return err
 		}
 		if pkg.ImportComment != "" {
-			importPath = pkg.ImportComment
+			// Work backwards to find the top-level import path
+			rel, err := filepath.Rel(src.Path, path)
+			if err != nil {
+				return errors.Wrapf(err, "Rel(%q, %q)", src.Path, path)
+			}
+
+			sub := filepath.ToSlash(rel)
+			p := pkg.ImportComment
+			if !strings.HasSuffix(p, sub) {
```

Comment:
I might be missing something, but what if `pkg.ImportComment` points to the same directory as `src.Path`, then this conditional will be true, causing `importPath` to remain undefined.

Reply:
This was a real bug, thanks. Added a test case and the fix.